### PR TITLE
WIP: use octosql.Value of protobuf type

### PR DESCRIPTION
**Pull request by cube2222**

### CODE REVIEW

File: values.go [hunk 1]
```diff
@@ -258,104 +181,298 @@ func NormalizeType(value interface{}) Value {
...
-	case Null:
-		_, ok := right.(Null)
-		if !ok {
-			return false
-		}
-		return true
+const (
+	LessThan    Comparison = -1
+	Equal       Comparison = 0
+	GreaterThan            = 1
```
Comment (woojtek) [inline values.go]: Comparison missing ?

File: values.go [hunk 2]
```diff
@@ -258,104 +181,298 @@ func NormalizeType(value interface{}) Value {
...
+func (v Value) AsFloat() float64 {
+	return v.GetFloat()
+}
+
+func (v Value) AsBool() bool {
+	return v.GetBool()
+}
+
+func (v Value) AsString() string {
+	return v.GetString_()
```
Comment (woojtek) [inline values.go]: '_' ?

File: values.go [hunk 3]
```diff
@@ -258,104 +181,298 @@ func NormalizeType(value interface{}) Value {
...
+
+	panic("unreachable")
+}
+
+func ZeroValue() Value {
+	return Value{}
+}
+
+func (v Value) AsInt() int {
+	return int(v.GetInt())
```
Comment (woojtek) [inline values.go]: why there (and only there) is type casting?

Comment (cube2222) [reply values.go]: Because getint is int64

File: execution/aggregates/min.go [hunk 4]
```diff
@@ -27,90 +27,27 @@ func (agg *Min) Document() docs.Documentation {
 	)
 }
 
-func (agg *Min) AddRecord(key octosql.Tuple, value octosql.Value) error {
+func (agg *Min) AddRecord(key octosql.Value, value octosql.Value) error {
```
[the remark this answers is not in the document]
Comment (cube2222) [reply execution/aggregates/min.go]: It's inside compare

File: execution/aggregates/max.go [hunk 5]
```diff
@@ -27,89 +27,27 @@ func (agg *Max) Document() docs.Documentation {
 	)
 }
 
-func (agg *Max) AddRecord(key octosql.Tuple, value octosql.Value) error {
+func (agg *Max) AddRecord(key octosql.Value, value octosql.Value) error {
```
[the remark this answers is not in the document]
Comment (cube2222) [reply execution/aggregates/max.go]: It's inside compare

File: execution/functions/functions.go [hunk 6]
```diff
@@ -154,19 +157,22 @@ var FuncAbs = execution.Function{
...
-		switch arg := args[0].(type) {
-		case Int:
-			if arg < 0 {
-				return -1 * arg, nil
+		arg := args[0]
+		switch arg.GetType() {
+		case TypeInt:
+			asInt := arg.AsInt()
+			if asInt < 0 {
+				return MakeInt(-1 * asInt), nil
```
[the remark this answers is not in the document]
Comment (cube2222) [reply execution/functions/functions.go]: we're not doing a comparison in negate, so asInt would only be used once.

File: execution/functions/functions.go [hunk 7]
```diff
@@ -154,19 +157,22 @@ var FuncAbs = execution.Function{
...
 			}
 			return arg, nil
-		case Float:
-			if arg < 0 {
-				return -1 * arg, nil
+		case TypeFloat:
+			asFloat := arg.AsFloat()
+			if asFloat < 0.0 {
+				return MakeFloat(-1.0 * asFloat), nil
```
[the remark this answers is not in the document]
Comment (cube2222) [reply execution/functions/functions.go]: we're not doing a comparison in negate, so asInt would only be used once.

File: values.go (hunk 1 above)
Comment (cube2222) [reply values.go]: Fixed

File: values.go (hunk 2 above)
Comment (cube2222) [reply values.go]: Protobuf quirk of the generated code.

File: execution/aggregates/max.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (cube2222) [reply execution/aggregates/max.go]: It's inside compare

**Pull request merged**